File: src/api/videos.ts
import { api } from "./client";
import type { Video } from "@/types";

interface ApiVideo {
  id: string;
  player_id: string;
  title: string;
  video_url?: string | null;
  thumbnail_url?: string | null;
  duration_sec?: number | null;
  skill_tag?: string | null;
  status: string;
  uploaded_at: string;
  rating?: number | null;
  assessment_date?: string | null;
  comment?: string | null;
  training_plan?: string | null;
  is_assessment?: boolean | null;
}

function getYouTubeThumbnail(url: string): string {
  const match = url.match(
    /(?:youtube\.com\/(?:watch\?v=|embed\/)|youtu\.be\/)([\w-]{11})/,
  );
  return match ? `https://img.youtube.com/vi/${match[1]}/hqdefault.jpg` : "";
}

function mapApiVideo(v: ApiVideo): Video {
  const mins = v.duration_sec ? Math.floor(v.duration_sec / 60) : 0;
  const secs = v.duration_sec ? v.duration_sec % 60 : 0;
  const videoUrl = v.video_url ?? undefined;
  const thumbnail =
    v.thumbnail_url ?? (videoUrl ? getYouTubeThumbnail(videoUrl) : "");
  return {
    id: v.id,
    playerId: v.player_id,
    title: v.title,
    thumbnail,
    duration: v.duration_sec
      ? `${mins}:${String(secs).padStart(2, "0")}`
      : "0:00",
    uploadedAt: v.uploaded_at,
    videoUrl,
    rating: v.rating ?? undefined,
    assessmentDate: v.assessment_date ?? undefined,
    comment: v.comment ?? undefined,
    trainingPlan: v.training_plan ?? undefined,
    isAssessment: v.is_assessment ?? undefined,
  };
}

export async function fetchPlayerVideos(playerId: string): Promise<Video[]> {
  const { data } = await api.get<ApiVideo[]>(`/api/v1/players/${playerId}/videos`);
  return data.map(mapApiVideo);
}

export async function uploadVideo(
  playerId: string,
  file: File,
  title: string,
  skillTag?: string,
): Promise<{ id: string }> {
  const formData = new FormData();
  formData.append("file", file);
  formData.append("title", title);
  if (skillTag) formData.append("skill_tag", skillTag);

  const { data } = await api.post(
    `/api/v1/players/${playerId}/videos`,
    formData,
    { headers: { "Content-Type": "multipart/form-data" } },
  );
  return data;
}

export async function getVideoUrl(videoId: string): Promise<string> {
  const { data } = await api.get<{ url: string }>(
    `/api/v1/videos/${videoId}/url`,
  );
  return data.url;
}

export async function addVideoLink(
  playerId: string,
  title: string,
  videoUrl: string,
  skillTag?: string,
): Promise<{ id: string }> {
  const { data } = await api.post(
    `/api/v1/players/${playerId}/video-links`,
    { title, video_url: videoUrl, skill_tag: skillTag },
  );
  return data;
}

export async function createAssessment(
  playerId: string,
  body: {
    title: string;
    video_url: string;
    rating?: number;
    comment?: string;
    training_plan?: string;
  },
): Promise<{ id: string; status: string; message: string }> {
  const { data } = await api.post(
    `/api/v1/players/${playerId}/assessments`,
    body,
  );
  return data;
}

File: backend/app/routers/videos.py
from datetime import date

from fastapi import APIRouter, Depends, UploadFile, File, Form, HTTPException, status
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.database import get_db
from app.models.user import User
from app.models.video import Video
from app.schemas.video import (
    VideoResponse,
    VideoUploadResponse,
    VideoLinkCreate,
    AssessmentCreate,
)
from app.services.auth_service import get_current_user
from app.services import player_service, video_service

router = APIRouter(prefix="/api/v1", tags=["videos"])

MAX_VIDEO_SIZE = 500 * 1024 * 1024  # 500MB


@router.post(
    "/players/{player_id}/videos",
    response_model=VideoUploadResponse,
    status_code=201,
)
async def upload_video(
    player_id: str,
    file: UploadFile = File(...),
    title: str = Form(...),
    skill_tag: str | None = Form(None),
    db: AsyncSession = Depends(get_db),
    user: User = Depends(get_current_user),
):
    """Upload a video for a player."""
    player = await player_service.get_player_or_404(db, player_id)
    player_service.assert_owner(player, user)

    if file.content_type not in ("video/mp4", "video/quicktime"):
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Допустимые форматы: mp4, mov",
        )

    content = await file.read()
    if len(content) > MAX_VIDEO_SIZE:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Файл превышает 500MB",
        )

    s3_key = video_service.upload_to_s3(content, file.filename or "video.mp4")

    video = Video(
        player_id=player_id,
        title=title,
        s3_key=s3_key,
        skill_tag=skill_tag,
        status="ready",
    )
    db.add(video)
    await db.commit()
    await db.refresh(video)

    return VideoUploadResponse(
        id=video.id, status=video.status, message="Видео загружено"
    )


@router.post(
    "/players/{player_id}/video-links",
    response_model=VideoUploadResponse,
    status_code=201,
)
async def add_video_link(
    player_id: str,
    body: VideoLinkCreate,
    db: AsyncSession = Depends(get_db),
    user: User = Depends(get_current_user),
):
    """Add a video by external link (YouTube, VK, direct mp4)."""
    player = await player_service.get_player_or_404(db, player_id)
    player_service.assert_owner(player, user)

    video = Video(
        player_id=player_id,
        title=body.title,
        video_url=body.video_url,
        skill_tag=body.skill_tag,
        status="ready",
    )
    db.add(video)
    await db.commit()
    await db.refresh(video)

    return VideoUploadResponse(
        id=video.id, status=video.status, message="Видео добавлено"
    )


@router.post(
    "/players/{player_id}/assessments",
    response_model=VideoUploadResponse,
    status_code=201,
)
async def create_assessment_video(
    player_id: str,
    body: AssessmentCreate,
    db: AsyncSession = Depends(get_db),
    user: User = Depends(get_current_user),
):
    player = await player_service.get_player_or_404(db, player_id)
    player_service.assert_owner(player, user)
    video = Video(
        player_id=player_id,
        title=body.title,
        video_url=body.video_url,
        rating=body.rating,
        comment=body.comment,
        training_plan=body.training_plan,
        is_assessment=True,
        assessment_date=date.today(),
        status="ready",
    )
    db.add(video)
    await db.commit()
    await db.refresh(video)
    return VideoUploadResponse(
        id=video.id, status=video.status, message="Оценка добавлена"
    )


@router.get("/players/{player_id}/videos", response_model=list[VideoResponse])
async def list_videos(player_id: str, db: AsyncSession = Depends(get_db)):
    """List all videos for a player."""
    result = await db.execute(
        select(Video)
        .where(Video.player_id == player_id)
        .order_by(Video.uploaded_at.desc())
    )
    videos = result.scalars().all()
    return [
        VideoResponse(
            id=v.id,
            player_id=v.player_id,
            title=v.title,
            video_url=v.video_url,
            thumbnail_url=None,
            duration_sec=v.duration_sec,
            skill_tag=v.skill_tag,
            status=v.status,
            uploaded_at=v.uploaded_at.isoformat(),
            rating=v.rating,
            assessment_date=v.assessment_date.isoformat()
            if v.assessment_date
            else None,
            comment=v.comment,
            training_plan=v.training_plan,
            is_assessment=v.is_assessment,
        )
        for v in videos
    ]


@router.get("/videos/{video_id}/url")
async def get_video_url(video_id: str, db: AsyncSession = Depends(get_db)):
    """Get a presigned URL for video playback (1h expiry)."""
    result = await db.execute(select(Video).where(Video.id == video_id))
    video = result.scalar_one_or_none()
    if not video:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Видео не найдено",
        )
    url = video_service.get_presigned_url(video.s3_key)
    return {"url": url, "expires_in": 3600}
